subscriber/common/sink/kafka.go

```diff
@@ -0,0 +1,181 @@
+		time.Duration(serviceConfig.ControllerConfig.Timeout)*time.Second,
+		http.Header{
+			"RPC-Caller":  []string{os.Getenv("UDEPLOY_APP_ID")},
+			"RPC-Service": []string{serviceConfig.ControllerConfig.ServiceName},
+		})
+	aresControllerClient.SetNamespace(cluster)
+
+	// replace httpSchemaFetcher with gateway client
+	// httpSchemaFetcher := NewHttpSchemaFetcher(httpClient, cfg.Address, metricScope)
+	cachedSchemaHandler := client.NewCachedSchemaHandler(
+		serviceConfig.Logger.Sugar(),
+		serviceConfig.Scope.Tagged(map[string]string{
+			"job":         jobConfig.Name,
+			"aresCluster": cluster,
+		}), aresControllerClient)
+
+	// schema refresh is based on job assignment refresh, so disable at here
+	err = cachedSchemaHandler.Start(0)
+	if err != nil {
+		return nil, err
+	}
+
+	kp := KafkaPublisher{
+		SyncProducer: p,
+		UpsertBatchBuilder: client.NewUpsertBatchBuilderImpl(
+			serviceConfig.Logger.Sugar(),
+			serviceConfig.Scope.Tagged(map[string]string{
+				"job":         jobConfig.Name,
+				"aresCluster": cluster,
+			}),
+			cachedSchemaHandler),
+		ServiceConfig: serviceConfig,
+		JobConfig:     jobConfig,
+		Scope: serviceConfig.Scope.Tagged(map[string]string{
+			"job":         jobConfig.Name,
+			"aresCluster": cluster,
+		}),
+		ClusterName: cluster,
+	}
+
+	return &kp, nil
+}
+
+// Shutdown will clean up resources that needs to be cleaned up
+func (kp *KafkaPublisher) Shutdown() {
+	kp.SyncProducer.Close()
+}
+
+// Save saves a batch of row objects into a destination
+func (kp *KafkaPublisher) Save(destination Destination, rows []client.Row) error {
+	shards := Sharding(rows, destination, kp.JobConfig)
+	if shards == nil {
+		// case1: no sharding --  publish rows to random kafka partition
+		kp.Insert(destination.Table, -1, destination.ColumnNames, rows, destination.AresUpdateModes...)
+	} else {
+		// case2: sharding -- publish rows to specified partition
+		for shardID, rowsInShard := range shards {
+			kp.Insert(destination.Table, int32(shardID), destination.ColumnNames, rowsInShard, destination.AresUpdateModes...)
+		}
+	}
+
+	return nil
+}
+
+// Cluster returns the DB cluster name
+func (kp *KafkaPublisher) Cluster() string {
+	return kp.ClusterName
+}
+
+func (kp *KafkaPublisher) Insert(tableName string, shardID int32, columnNames []string, rows []client.Row,
+	updateModes ...memCom.ColumnUpdateMode) (int, error) {
+	kp.Scope.Gauge("batchSize").Update(float64(len(rows)))
+	saveStart := time.Now()
+	kp.ServiceConfig.Logger.Debug("saving", zap.Any("rows", rows))
+
+	bytes, numRows, err := kp.UpsertBatchBuilder.PrepareUpsertBatch(tableName, columnNames, updateModes, rows)
+	if err != nil {
+		kp.Scope.Counter("errors.insert").Inc(1)
+		return 0, utils.StackError(err, fmt.Sprintf("Failed to prepare rows in table %s, columns: %+v",
+			tableName, columnNames))
+	}
+
+	msg := sarama.ProducerMessage{
+		Topic: fmt.Sprintf("%s-%s", tableName, kp.Cluster()),
```

Comment:
nit: fix to namespace first, then table name?

Reply:
done